### PR TITLE
organize by model; update with new models definitions

### PR DESCRIPTION
**Pull request by rshewitt**
# Pull Request

Related to [#4757](https://github.com/GSA/data.gov/issues/4757)

## About

update model definitions and use of models 

## PR TASKS

- [x] The actual code changes.
- [x] Tests written and passed.
- [ ] Any changes to docs?
- [x] Bumped version number in [pyproject.toml](https://github.com/GSA/datagov-harvesting-logic/blob/8078c5b9f015ca7fb8a142e4e4b2e22ad2fd49b1/pyproject.toml#L3) (also checked on [PyPi](https://pypi.org/project/ckanext-geodatagov/#history)).

### CODE REVIEW

File: database/interface.py [hunk 1]
```diff
@@ -199,19 +209,26 @@ def add_harvest_error(self, error_data):
             self.db.rollback()
             return None
 
-    def get_harvest_error(self, error_id):
-        result = self.db.query(HarvestError).filter_by(id=error_id).first()
+    def get_harvest_job_error(self, job_id: str) -> dict:
```
Comment (btylerburton) [inline database/interface.py:224]: now that we have HarvestJobError and HarvestRecordError, and we will likely never have both in a single job, we could just query both tables with a job_id to remain consistent.

as there is definitely a case for all errors associated with a job_id for reporting results of a harvest run.

File: app/routes.py [hunk 2]
```diff
@@ -265,6 +251,15 @@ def get_harvest_job(job_id=None):
         return "Please provide correct job_id or harvest_source_id"
 
 
+@mod.route("/harvest_job/<job_id>/error", methods=["GET"])
+def get_harvest_job_error(job_id):
+    try:
+        job_error = db.get_harvest_job_error(job_id)
+        return jsonify(job_error) if job_id else ("Not Found", 404)
```
Comment (btylerburton) [inline app/routes.py]: shouldn't do the query if we don't have job_id

File: app/routes.py [hunk 3]
```diff
@@ -352,6 +312,36 @@ def get_harvest_record(record_id=None):
...
+@mod.route("/harvest_record/<record_id>/errors", methods=["GET"])
+def get_all_harvest_record_errors(record_id: str) -> list:
+    try:
+        record_errors = db.get_harvest_record_errors(record_id)
+        return record_errors if record_errors else ("Not Found", 404)
+    except Exception:
+        return "Please provide correct record_id"
+
+
+@mod.route("/harvest_record/<error_id>/error", methods=["GET"])
```
Comment (btylerburton) [inline app/routes.py]: if passing `error_id` i'd think route would be `harvest_error/<error_id>` and then we'd want to query both Error tables.

File: database/interface.py [hunk 4]
```diff
@@ -187,9 +194,12 @@ def delete_harvest_job(self, job_id):
         self.db.commit()
         return "Harvest job deleted successfully"
 
-    def add_harvest_error(self, error_data):
+    def add_harvest_error(self, error_data: dict, error_type: str):
```
Comment (btylerburton) [inline database/interface.py:197]: probably want a specific response for not passing `error_type` as that's going to be easy enough to forget now.

File: database/models.py [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
 from flask_sqlalchemy import SQLAlchemy
-from sqlalchemy import Enum, func
+from sqlalchemy import Enum, func, String, Column
+from sqlalchemy.orm import DeclarativeBase
 
-db = SQLAlchemy()
 
-
-class Base(db.Model):
+class Base(DeclarativeBase):
```
Comment (btylerburton) [inline database/models.py:8]: what's the advantage of shifting to this as base inheritance class?

File: database/models.py [hunk 6]
```diff
@@ -1,17 +1,26 @@
...
     __abstract__ = True  # Indicates that this class should not be created as a table
-    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
+    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
+
+
+db = SQLAlchemy(model_class=Base)
 
 
-class Organization(Base):
+class Error(db.Model):
```
Comment (btylerburton) [inline database/models.py:16]: we should include `harvest_job_id` as common to both errors as we will need to correlate record level errors to the job unless i'm missing something.

**Review by btylerburton — CHANGES_REQUESTED** [review]
This complicates record level error lookups by jobid which we'll need for harvest reporting.

File: database/interface.py (hunk 1 above)
Comment (rshewitt) [reply database/interface.py:224]: add [here](https://github.com/GSA/datagov-harvesting-logic/blob/8baa21616de9693b43ff78e40f8e05d0cfeba56e/database/interface.py#L214)

File: app/routes.py (hunk 2 above)
Comment (rshewitt) [reply app/routes.py]: paired and discussed

File: database/interface.py (hunk 4 above)
Comment (rshewitt) [reply database/interface.py:197]: added [here](https://github.com/GSA/datagov-harvesting-logic/blob/8baa21616de9693b43ff78e40f8e05d0cfeba56e/database/interface.py#L199)

File: database/models.py (hunk 5 above)
Comment (rshewitt) [reply database/models.py:8]: paired and discussed

File: database/models.py (hunk 6 above)
Comment (rshewitt) [reply database/models.py:16]: [resolved with](https://github.com/GSA/datagov-harvesting-logic/blob/8baa21616de9693b43ff78e40f8e05d0cfeba56e/database/interface.py#L214)

File: app/routes.py [hunk 7]
```diff
@@ -352,6 +319,36 @@ def get_harvest_record(record_id=None):
...
```
Comment (btylerburton) [inline app/routes.py]: shouldn't this be `harvest_error/<error_id>` for consistency?

File: tests/integration/database/test_db.py [hunk 8]
```diff
@@ -101,6 +102,43 @@ def test_harvest_source_by_jobid(
...
+        source_data_dcatus,
+        job_data_dcatus,
+        job_error_data,
+    ):
+        interface.add_organization(organization_data)
+        interface.add_harvest_source(source_data_dcatus)
+        interface.add_harvest_job(job_data_dcatus)
+
+        harvest_job_error = interface.add_harvest_error(job_error_data, "job")
+        assert isinstance(harvest_job_error, HarvestJobError)
```
Comment (btylerburton) [inline tests/integration/database/test_db.py:118]: nice assertion!

**Review by btylerburton — DISMISSED** [review]
Great cleanup!

File: app/routes.py (hunk 7 above)
Comment (rshewitt) [reply app/routes.py]: [resolved](https://github.com/GSA/datagov-harvesting-logic/pull/80/commits/6e745b6ccf238eb8d42e37cf3aff6248672aa2a5)

**Review by btylerburton — APPROVED** [review]

**Pull request merged**